homeassistant/components/emulated_kasa/__init__.py

```diff
@@ -0,0 +1,112 @@
+"""Support for local power state reporting of entities by emulating TP-Link Kasa smart plugs."""
+import logging
+
+from sense_energy import PlugInstance, SenseLink
+import voluptuous as vol
+
+from homeassistant.components.switch import ATTR_CURRENT_POWER_W
+from homeassistant.const import (
+    CONF_ENTITIES,
+    CONF_NAME,
+    EVENT_HOMEASSISTANT_STARTED,
+    EVENT_HOMEASSISTANT_STOP,
+    STATE_ON,
+)
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.template import Template, is_template_string
+
+from .const import CONF_POWER, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_ENTITY_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(CONF_POWER): vol.Any(
+            vol.Coerce(float),
+            cv.template,
+        ),
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_ENTITIES): vol.Schema(
+                    {cv.entity_id: CONFIG_ENTITY_SCHEMA}
+                ),
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Sense component."""
+    hass.data.setdefault(DOMAIN, {})
+    conf = config.get(DOMAIN)
+    if not conf:
+        return True
+    entity_configs = conf.get(CONF_ENTITIES, {})
+
+    def devices():
+        """Devices to be emulated."""
+        yield from get_plug_devices(hass, entity_configs)
+
+    server = SenseLink(devices)
+
+    async def stop_emulated_kasa(event):
+        await server.stop()
+
+    async def start_emulated_kasa(event):
+        for entity_id, entity_config in entity_configs.items():
+            state = hass.states.get(entity_id)
```

Comment:
It would be cleaner to create the Template here and set `power_val.hass = hass` so that only has to happen once